Add function-like behavior for Taylor1, TaylorN, HomogeneousPolynomial

PerezHz opened the pull request:
This PR implements a solution for #116, and adds corresponding tests; tests are passing on my fork, thanks everyone for your feedback!

PerezHz commented:
PR tests are passing, if there is anything that I missed or that you think should be added, please let me know 😄

blas-ko commented:
> if there is anything that I missed or that you think should be added, please let me know.

It would be a nice feature to take advantage of this PR to implement an evaluate method for an array of values. I'm thinking of something like
```julia
julia> using TaylorSeries
julia> import TaylorSeries: evaluate;
julia> function evaluate(p::Taylor1{T},x::Vector{S}) where {T<:Number,S<:Number}
    ll = length(x)
    p_eval = zeros(x)
    for (i,xi) in enumerate(x)
        @inbounds p_eval[i] = evaluate(p,xi)
    end
    return p_eval
end
julia> (p::Taylor1)(x) = evaluate(p,x) #this now works for numbers & arrays!!
julia> p = Taylor1([1,2,3,4])
 1 + 2 t + 3 t² + 4 t³ + 𝒪(t⁴)
julia> p(1.0)
 10.0
julia> p([1.0,0.0])
2-element Array{Float64,1}:
 10.0
  1.0
```
The analogous could be done for `TaylorN` but instead an `NxM` matrix should be passed as an argument where `N`: number of variables, `M`: number of values to be evaluated.

PerezHz commented:
I agree that this functionality should be available for the cases you mention, @blas-ko! Granted, I didn't add the methods you are proposing; but actually, as the way it is now, such cases may already be handled via broadcasting:

```julia
#evaluating p at a matrix of Float64s:
julia> p.([1.0 0.0; -1.0 -2.0])
2×2 Array{Float64,2}:
 10.0    1.0
 -2.0  -23.0

#evaluating p at a 3x3x3 matrix of Float64s:
julia> p.(rand(3,3,3))
3×3×3 Array{Float64,3}:
[:, :, 1] =
 8.65957  1.17927  8.12569
 1.13276  2.57914  3.37285
 2.51999  5.42389  2.04953

[:, :, 2] =
 1.03394  1.19053  3.51668
 1.15536  9.98952  4.27719
 7.29441  2.3461   2.76938

[:, :, 3] =
 1.33343  1.11886  1.08465
 5.75461  1.47925  1.78689
 1.47734  2.19666  1.84005
```
and for TaylorNs:
```julia
julia> ξ1, ξ2 = set_variables("ξ", numvars=2, order=15);

julia> f = sin(ξ1+ξ1*ξ2); g = exp(-sin(im*ξ1)+ξ1*ξ2); h = [f, g];

julia> f = sin(ξ1+ξ1*ξ2); g = exp(-sin(ξ1)+ξ1*ξ2); h = [f, g];

#evaluating f. at a vector of evaluation points returns a vector of Float64 values:
julia> f.([rand(2),rand(2)])
2-element Array{Float64,1}:
 0.974375
 0.585455

#evaluating g. at a 2x2 matrix of evaluation points returns a 2x2 matrix of Float64 values:
julia> g.( map(x->rand(2), rand(2,2)) )
2×2 Array{Float64,2}:
 0.929124  0.932631
 0.825413  0.888743

#evaluating h=[f,g] at a 2x2 matrix of evaluation points returns a 2x2 matrix of Vector{Float64}(2) values:
julia> h.( map(x->rand(2), rand(2,2)) )
2×2 Array{Array{Float64,1},2}:
 [0.674398, 0.736717]  [0.991339, 0.994949]
 [0.865741, 0.959931]  [0.460701, 0.800402]

#and so on...
julia> h.( [rand(2),rand(2)] )
2-element Array{Array{Float64,1},1}:
 [0.748772, 0.728316]
 [0.165423, 0.964655]

julia> h.( map(x->rand(2), rand(3,2,2)) )
3×2×2 Array{Array{Float64,1},3}:
[:, :, 1] =
 [0.970154, 0.996443]  [0.952156, 0.628598]
 [0.434873, 0.979518]  [0.750666, 0.492932]
 [0.839201, 0.475184]  [0.151065, 0.894432]

[:, :, 2] =
 [0.966398, 0.597349]  [0.414589, 0.901062]
 [0.045531, 0.992213]  [0.58178, 0.701201] 
 [0.395325, 0.718587]  [0.951338, 0.66144] 
```

I think this is, in a sense, more julian, but also more general; otherwise one would have to define a method for each particular case... The only caveat is that this behavior is only supported by julia 0.6+

PerezHz commented:
So, for the particular case case of Taylor1 it might be enough to define:

```julia
evaluate{T<:Number,S<:Number}(p::Taylor1{T},x::Vector{S}) = evaluate.(p,x)
```

Thus

```julia
julia> p = Taylor1(rand(10))

julia> vr = rand(5)

julia> p(vr) == p.(vr)
true

julia> p(vr) == evaluate.(p,vr)
true
```

I'm gonna push this extra method for Taylor1s and corresponding tests! Thanks @blas-ko for your suggestion! I'd like to use some similar broadcasting tricks for TaylorN, but then the second argument would have to be a `Vector{Vector{S}}`...

PerezHz commented:
Is there an easy way to convert from a `Vector{Vector{T}}` to an `Array{T,2}`?

blas-ko commented:
>  such cases may already be handled via broadcasting

You're absolutely right, my fault there!

> Is there an easy way to convert from a Vector{Vector{T}} to an Array{T,2}?

Maybe `hcat`
```julia
julia> a = [[1,2,3],[4,5,6]];
julia> @time hcat(a...)
  0.000038 seconds (8 allocations: 384 bytes)
3×2 Array{Int64,2}:
 1  1
 1  1
 2  3
```

PerezHz commented:
Many thanks, @blas-ko!

PerezHz commented:
> Is everything type stable? It seems to me that this is the case...

Well, as this functionality is being proposed right now, the function-like behavior of Taylor1, etc., at the end of the day relies on `evaluate`, so in principle this new behavior should be as type-stable as `evaluate` is... Do you have any concrete cases in mind?

> What about mixed-types, i.e. Taylor1{TaylorN{T}}, or TaylorN{Taylor1{T}} ?

I'm gonna add some tests for the mixtures! That's actually something important that I didn't add! 😅 sorry about that!  Regarding the function-like behavior for mixtures, this again relies on `evaluate`, so there should be no problem, but I'm adding some tests which should check this

PerezHz commented:
Thanks for all your comments @lbenet! I actually had to tweak a bit some parts of `src/evaluate.jl` for the mixtures, but now everything's working smoothly. Tests are passing locally on julia 0.5.2, 0.6.0 and 0.7.0-DEV.1486, but let's see what travis thinks... An important thing that I would like to note is that the tweaks in `src/evaluate.jl` proposed here are also consistent with TaylorIntegration.jl tests on latest master for julia 0.5.2, 0.6.0 and 0.7.0-DEV

PerezHz commented:
@lbenet @dpsanders there was a weird error on travis [job#652.5](https://travis-ci.org/JuliaDiff/TaylorSeries.jl/jobs/270596125), could you please help me restart it? otherwise, tests are passing

dpsanders commented:
I restarted the job. 

Actually, with the syntax t(3) to evaluate, is there now any need for the function called evaluate?

PerezHz commented:
>I restarted the job.

Thanks, David!

> Actually, with the syntax t(3) to evaluate, is there now any need for the function called evaluate?

Well, currently `t(3)` is only an alternative syntax for `evaluate(t,3)` (is that what is called syntactic sugar?), i.e. `t(3) == evaluate(t,3)`, where t is a Taylor1... But I agree, if done carefully, I think it would be possible to get rid completely of the `evaluate` function, if we substituted the code for the evaluate methods in the definition of the `(p::Taylor1)(x)` methods, etc. On the other hand, probably there still would be a need for the in-place version of `evaluate` (i.e., the `evaluate!` function).

PerezHz commented:
The last commit addresses more mixture cases, such as evaluating a `TaylorN{Taylor1{Float64}}` or an `Array{TaylorN{Taylor1{Float64}}}` at an `Array{Taylor1{Float64}}`; I added the corresponding `evaluate` methods and tests... On travis, once again there was a job ([travis job #653.5](https://travis-ci.org/JuliaDiff/TaylorSeries.jl/jobs/270718728)) with some weird errors which had nothing to do with TaylorSeries.jl; after restarting everything should be working fine... Locally, tests are passing in julia 0.5.2, 0.6.0 and 0.7.0-DEV.1599

lbenet commented:
> Actually, with the syntax t(3) to evaluate, is there now any need for the function called evaluate?

Probably we should keep `evaluate` for a while, and eventually not exporting it or simply deprecate it.

I just restarted one test that hung.

PerezHz commented:
> Probably we should keep evaluate for a while, and eventually not exporting it or simply deprecate it.

I agree!

> I just restarted one test that hung.

Many thanks, @lbenet!

lbenet commented:
Tests are passing. Could you add something to the docs, and then we merge this? Thanks for all the work!

PerezHz commented:
> Could you add something to the docs, and then we merge this?

Sure! I added some general details about this new behavior in the User's Guide section of the docs. I was also thinking that I should add some docstrings, but I was thinking whether I should add those in the  `evaluate` docstrings, or in the docstrings for the constructors of `Taylor1`, `HomogeneousPolynomial` and `TaylorN` variables?

I'm asking this because, as I understood from JuliaDocs/Documenter.jl#228 and JuliaDocs/Documenter.jl#290, currently it is not possible to write docstrings for these kinds of methods, and the recommended approach mentioned there is to document within the type's docstring that objects of that type are callable and provide a usage example there. So, how would you recommend me to proceed?

PerezHz commented:
I've updated the docstrings for `evaluate` as well as the docstrings for the `Taylor1`, `HomogeneousPolynomial` and `TaylorN` constructors. That's the only change since last commit, so travis should give green lights. If there is anything that you think should be added, please let me know; otherwise, this should be ready to be merged...

PerezHz commented:
Job `#19.4` is failing on travis... While working with TaylorSeries.jl, I've seen this error is related to a test which fails from time to time (`test/onevariable.jl`, line 343) due to use of random variables, but if re-run, then it should pass

PerezHz commented:
Tests are now passing!

lbenet commented on src/evaluate.jl:
```diff
@@ -88,8 +97,25 @@ function evaluate{T<:Number}(a::Taylor1{T}, x::Taylor1{T})
     end
     suma
 end
+function evaluate{T<:NumberNotSeries}(a::Taylor1{Taylor1{T}}, x::Taylor1{T})
```
`NumberNotSeries` here may be too restrictive; I'm thinking in cases such as `Taylor1{Taylor1{...{T}...}`. Would everything work if `T<:Number`?

PerezHz replied:
Sure, it should work! Thanks!

lbenet commented on src/evaluate.jl:
```diff
@@ -100,6 +126,25 @@ function evaluate!{T<:Number}(x::Array{TaylorN{T},1}, δx::Array{T,1},
     end
     nothing
 end
+
```
I think the two methods below can be written as one, imposing `T<:NumberNotSeriesN`.

PerezHz replied:
Thanks @lbenet!

lbenet commented on src/evaluate.jl:
```diff
@@ -190,10 +245,59 @@ function evaluate{T<:Number,S<:NumberNotSeries}(a::TaylorN{T},
     return x0
 end
+
+function evaluate{T<:NumberNotSeries}(x::Array{TaylorN{T},1}, δx::Array{Taylor1{T},1})
```
Again, I think the two following methods can be written as one, using `T<:NumberNotSeriesN`.

PerezHz replied:
👌Great! Thanks!

PerezHz commented:
@lbenet: just pushed the changes you suggested; tests are passing locally on 0.5.2, 0.6.0, and 0.7.0-DEV and are running on travis right now. Again, many thanks!

lbenet commented:
Thanks a lot for your contribution! I am merging it right now!

blas-ko commented:
I was just checking evaluation and it seems that broadcasting is pretty slow; check the following
```julia
julia> using TaylorSeries; import TaylorSeries: evaluate 

julia> function evaluate{T<:Number}(a::Taylor1{T},vals::Vector{T})
           ll = length(vals)
           sumas = zeros(T,ll)
           for i in 1:ll
               sumas[i] = evaluate(a,vals[i])
           end
           return sumas
       end
evaluate (generic function with 15 methods)

julia> function evaluate{T<:Number,S<:Number}(a::Taylor1{T},vals::Vector{S})
           ll = length(vals)
           R = promote_type(S,T)
           sumas = zeros(R,ll)
           for i in 1:ll
               sumas[i] = evaluate(a,vals[i])
           end
           return sumas
       end
evaluate (generic function with 16 methods)

julia> t = Taylor1(10); a = collect(0.:0.5:10);

julia> using BenchmarkTools

julia> @btime evaluate(t,a); @btime evaluate.(t,a);
  402.035 ns (1 allocation: 256 bytes)
  7.380 μs (26 allocations: 1.31 KiB)
```
I've not done this with the other cases, but in this particular case the difference is _yuuuge_. Maybe it's worth it to write the specific methods for evaluating in vectors or matrices.

lbenet commented:
@blas-ko I just merged this PR, seemingly at the same time you wrote your comment, though I was a bit faster ;-).

You are right that broadcasting (here) incurs in some allocations that cost time, roughly a factor 18. Note that
```julia
julia> @which evaluate.(t,a)
broadcast(f, A, Bs...) in Base.Broadcast at broadcast.jl:434
```
so my guess is that internally `broadcast` calls `evaluate(t,a[i])` with some internal evaluations,  which are not as efficient as the explicit method. In other words, if there is a method that does broadcasting, use it.
